gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+            let input_bytes = input.hash || input.index
+            raw_items.append(input_bytes)
+
+    let N = len(raw_items)
+    let F = N * P
+
+    let hashed_items = []
+    for raw_item in raw_items:
+        let hashed_item = (siphash_key(siphash_key, raw_item) * F) >> 64
+        hashed_items.append(hashed_item)
+
+    hashed_items.sort()
+
+    gcs_compress(hashed_items, fp)
+</pre>
+
+
+Given a Bitcoin block, a full-node construct an <code>Extended</code> compact filter as follows
+<pre>
+construct_extended_gcs_filter(block, fp) -> []byte:
+
+    let siphash_key = block.hash()[:16]
+
+    let P = 1 << fp
+
+    let raw_items = []
+    for tx in block.transactions:
+        if tx.is_coinbase():
+           continue
+
+        for input in tx.inputs:
+            for wit_elem in input.witness:
+                raw_items.append(wit_elem)
+
+            let sig_script_pushes = extract_push_datas(input.sig_script)
+            for push in sig_script_pushes:
+                raw_items.append(push)
+
+    let N = len(raw_items)
+    let F = N * P
+
+    let hashed_items = []
+    for raw_item in raw_items:
+        let hashed_item = (siphash_key(siphash_key, raw_item) * F) >> 64
+        hashed_items.append(hashed_item)
+
+    hashed_items.sort()
+
+    gcs_compress(hashed_items, fp)
+</pre>
+
+==== Filter Capability Querying ====
+
+As it's feasible that in the future, this document is extended to encompass
+additional filter encoding algorithms or filter contents, we define a new p2p
+message that allows light clients to ascertain which filters a node supports.
+
+The <code>getcftypes</code> message is an ''empty message'' whose command string is:
+<code>getcftypes</code>
+
+A full-node that receives a <code>getcftypes</code> message should respond with a
+<code>cftypes</code> message which is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| Var-Int
+| NumFilters
+| uint64
+| The number of supported filters.
+|-
+| NumFilters
+| SupportedFilters
+| [NumFilterBytes]byte
+| A byte slice with each byte denoting a supported filter type
+|}
+
+
+==== Compact Filter Header Chain ====
+
+As the filters described in this BIP ''are not'' consensus critical, meaning
+each filter is validated by full-nodes and committed into blocks by miners, we
+require an alternative (albeit less-binding) method to allow light clients to
+''identify'' and ''reject'' invalid filters. The purely p2p solution to this
+problem is to obtain a deterministic hash-chain of ''each'' filter. This hash
+chain or "filter header chain" is similar to the regular Bitcoin headers in
+that it allows a light client to verify the ''authenticity'' of a received
+filter.
+
+The filter header chain for a particular filter type is described by the
+following recurrence:
+<pre>
+filter_header(n: uint) -> [32]byte = 
+   let zero_hash [32]byte = {0*32}
+
+   if n == 0:
+       double-sha-256(genesis_block.prevblock || filter(0))
```

Comment:
Soooo...0? Please just say 32 bytes of 0s.

Reply:
Fixed.